File: src/main/java/net/green_jab/fixed_minecraft/block/custom/NetheriteAnvilBlock.java
package net.green_jab.fixed_minecraft.block.custom;

import com.mojang.serialization.MapCodec;
import net.green_jab.fixed_minecraft.block.ModBlocks;
import net.minecraft.block.*;
import net.minecraft.entity.Entity;
import net.minecraft.entity.FallingBlockEntity;
import net.minecraft.entity.ai.pathing.NavigationType;
import net.minecraft.entity.damage.DamageSource;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemPlacementContext;
import net.minecraft.screen.AnvilScreenHandler;
import net.minecraft.screen.NamedScreenHandlerFactory;
import net.minecraft.screen.ScreenHandlerContext;
import net.minecraft.screen.SimpleNamedScreenHandlerFactory;
import net.minecraft.stat.Stats;
import net.minecraft.state.StateManager;
import net.minecraft.state.property.DirectionProperty;
import net.minecraft.text.Text;
import net.minecraft.util.ActionResult;
import net.minecraft.util.BlockRotation;
import net.minecraft.util.Hand;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.shape.VoxelShape;
import net.minecraft.util.shape.VoxelShapes;
import net.minecraft.world.BlockView;
import net.minecraft.world.World;
import net.minecraft.world.WorldEvents;
import org.jetbrains.annotations.Nullable;

public class NetheriteAnvilBlock extends FallingBlock {
    public static final MapCodec<NetheriteAnvilBlock> CODEC = NetheriteAnvilBlock.createCodec(NetheriteAnvilBlock::new);
    public static final DirectionProperty FACING = HorizontalFacingBlock.FACING;
    private static final VoxelShape BASE_SHAPE = Block.createCuboidShape(2.0, 0.0, 2.0, 14.0, 4.0, 14.0);
    private static final VoxelShape X_STEP_SHAPE = Block.createCuboidShape(3.0, 4.0, 4.0, 13.0, 5.0, 12.0);
    private static final VoxelShape X_STEM_SHAPE = Block.createCuboidShape(4.0, 5.0, 6.0, 12.0, 10.0, 10.0);
    private static final VoxelShape X_FACE_SHAPE = Block.createCuboidShape(0.0, 10.0, 3.0, 16.0, 16.0, 13.0);
    private static final VoxelShape Z_STEP_SHAPE = Block.createCuboidShape(4.0, 4.0, 3.0, 12.0, 5.0, 13.0);
    private static final VoxelShape Z_STEM_SHAPE = Block.createCuboidShape(6.0, 5.0, 4.0, 10.0, 10.0, 12.0);
    private static final VoxelShape Z_FACE_SHAPE = Block.createCuboidShape(3.0, 10.0, 0.0, 13.0, 16.0, 16.0);
    private static final VoxelShape X_AXIS_SHAPE = VoxelShapes.union(BASE_SHAPE, X_STEP_SHAPE, X_STEM_SHAPE, X_FACE_SHAPE);
    private static final VoxelShape Z_AXIS_SHAPE = VoxelShapes.union(BASE_SHAPE, Z_STEP_SHAPE, Z_STEM_SHAPE, Z_FACE_SHAPE);
    private static final Text TITLE = Text.translatable("container.repair");
    private static final float FALLING_BLOCK_ENTITY_DAMAGE_MULTIPLIER = 2.0f;
    private static final int FALLING_BLOCK_ENTITY_MAX_DAMAGE = 40;

    public MapCodec<NetheriteAnvilBlock> getCodec() {
        return CODEC;
    }

    public NetheriteAnvilBlock(AbstractBlock.Settings settings) {
        super(settings);
        this.setDefaultState((BlockState)((BlockState)this.stateManager.getDefaultState()).with(FACING, Direction.NORTH));
    }

    @Override
    public BlockState getPlacementState(ItemPlacementContext ctx) {
        return (BlockState)this.getDefaultState().with(FACING, ctx.getHorizontalPlayerFacing().rotateYClockwise());
    }

    @Override
    public ActionResult onUse(BlockState state, World world, BlockPos pos, PlayerEntity player, Hand hand, BlockHitResult hit) {
        if (world.isClient) {
            return ActionResult.SUCCESS;
        }
        player.openHandledScreen(state.createScreenHandlerFactory(world, pos));
        player.incrementStat(Stats.INTERACT_WITH_ANVIL);
        return ActionResult.CONSUME;
    }

    @Override
    @Nullable
    public NamedScreenHandlerFactory createScreenHandlerFactory(BlockState state, World world, BlockPos pos) {
        return new SimpleNamedScreenHandlerFactory((syncId, inventory, player) -> new AnvilScreenHandler(syncId, inventory, ScreenHandlerContext.create(world, pos)), TITLE);
    }

    @Override
    public VoxelShape getOutlineShape(BlockState state, BlockView world, BlockPos pos, ShapeContext context) {
        Direction direction = state.get(FACING);
        if (direction.getAxis() == Direction.Axis.X) {
            return X_AXIS_SHAPE;
        }
        return Z_AXIS_SHAPE;
    }

    @Override
    protected void configureFallingBlockEntity(FallingBlockEntity entity) {
        entity.setHurtEntities(2.0f, 40);
    }

    @Override
    public void onLanding(World world, BlockPos pos, BlockState fallingBlockState, BlockState currentStateInPos, FallingBlockEntity fallingBlockEntity) {
        if (!fallingBlockEntity.isSilent()) {
            world.syncWorldEvent(WorldEvents.ANVIL_LANDS, pos, 0);
        }
    }

    @Override
    public void onDestroyedOnLanding(World world, BlockPos pos, FallingBlockEntity fallingBlockEntity) {
        if (!fallingBlockEntity.isSilent()) {
            world.syncWorldEvent(WorldEvents.ANVIL_DESTROYED, pos, 0);
        }
    }

    @Override
    public DamageSource getDamageSource(Entity attacker) {
        return attacker.getDamageSources().fallingAnvil(attacker);
    }

    @Nullable
    public static BlockState getLandingState(BlockState fallingState) {
        if (fallingState.isOf(ModBlocks.NETHERITE_ANVIL)) {
            return (BlockState)ModBlocks.CHIPPED_NETHERITE_ANVIL.getDefaultState().with(FACING, fallingState.get(FACING));
        }
        if (fallingState.isOf(ModBlocks.CHIPPED_NETHERITE_ANVIL)) {
            return (BlockState)ModBlocks.DAMAGED_NETHERITE_ANVIL.getDefaultState().with(FACING, fallingState.get(FACING));
        }
        return null;
    }

    @Override
    public BlockState rotate(BlockState state, BlockRotation rotation) {
        return (BlockState)state.with(FACING, rotation.rotate(state.get(FACING)));
    }

    @Override
    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
        builder.add(FACING);
    }

    @Override
    public boolean canPathfindThrough(BlockState state, BlockView world, BlockPos pos, NavigationType type) {
        return false;
    }

    @Override
    public int getColor(BlockState state, BlockView world, BlockPos pos) {
        return state.getMapColor((BlockView)world, (BlockPos)pos).color;
    }
}

File: src/main/kotlin/net/greenjab/fixedminecraft/blocks/anvil/NetheriteAnvilBlock.kt
package net.greenjab.fixedminecraft.blocks.anvil

import com.mojang.serialization.MapCodec
import net.greenjab.fixedminecraft.blocks.BlockRegistry
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.FallingBlock
import net.minecraft.block.HorizontalFacingBlock
import net.minecraft.block.ShapeContext
import net.minecraft.entity.Entity
import net.minecraft.entity.FallingBlockEntity
import net.minecraft.entity.ai.pathing.NavigationType
import net.minecraft.entity.damage.DamageSource
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemPlacementContext
import net.minecraft.screen.AnvilScreenHandler
import net.minecraft.screen.NamedScreenHandlerFactory
import net.minecraft.screen.ScreenHandlerContext
import net.minecraft.screen.SimpleNamedScreenHandlerFactory
import net.minecraft.stat.Stats
import net.minecraft.state.StateManager
import net.minecraft.state.property.DirectionProperty
import net.minecraft.text.Text
import net.minecraft.util.ActionResult
import net.minecraft.util.BlockRotation
import net.minecraft.util.Hand
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.shape.VoxelShape
import net.minecraft.util.shape.VoxelShapes
import net.minecraft.world.BlockView
import net.minecraft.world.World
import net.minecraft.world.WorldEvents

class NetheriteAnvilBlock(settings: Settings) : FallingBlock(settings) {
    public override fun getCodec() = CODEC

    init {
        defaultState = stateManager.defaultState.with(FACING, Direction.NORTH)
    }

    override fun getPlacementState(ctx: ItemPlacementContext): BlockState {
        return defaultState.with(FACING, ctx.horizontalPlayerFacing.rotateYClockwise())
    }

    override fun onUse(
        state: BlockState,
        world: World,
        pos: BlockPos,
        player: PlayerEntity,
        hand: Hand,
        hit: BlockHitResult,
    ): ActionResult {
        // hmmm
        if (world.isClient) {
            return ActionResult.SUCCESS
        }
        player.openHandledScreen(state.createScreenHandlerFactory(world, pos))
        player.incrementStat(Stats.INTERACT_WITH_ANVIL)
        return ActionResult.CONSUME
    }

    override fun createScreenHandlerFactory(state: BlockState, world: World, pos: BlockPos): NamedScreenHandlerFactory {
        return SimpleNamedScreenHandlerFactory({ syncId, inventory, _ ->
            AnvilScreenHandler(syncId, inventory, ScreenHandlerContext.create(world, pos))
        }, TITLE)
    }

    override fun getOutlineShape(state: BlockState, world: BlockView, pos: BlockPos, context: ShapeContext): VoxelShape {
        val direction = state.get(FACING)
        return if (direction.axis === Direction.Axis.X)
            X_AXIS_SHAPE
        else
            Z_AXIS_SHAPE
    }

    override fun configureFallingBlockEntity(entity: FallingBlockEntity) {
        entity.setHurtEntities(FALLING_BLOCK_ENTITY_DAMAGE_MULTIPLIER, FALLING_BLOCK_ENTITY_MAX_DAMAGE)
    }

    override fun onLanding(
        world: World,
        pos: BlockPos,
        fallingBlockState: BlockState,
        currentStateInPos: BlockState,
        fallingBlockEntity: FallingBlockEntity,
    ) {
        if (!fallingBlockEntity.isSilent)
            world.syncWorldEvent(WorldEvents.ANVIL_LANDS, pos, 0)
    }

    override fun onDestroyedOnLanding(world: World, pos: BlockPos, fallingBlockEntity: FallingBlockEntity) {
        if (!fallingBlockEntity.isSilent)
            world.syncWorldEvent(WorldEvents.ANVIL_DESTROYED, pos, 0)
    }

    override fun getDamageSource(attacker: Entity): DamageSource = attacker.damageSources.fallingAnvil(attacker)

    override fun rotate(state: BlockState, rotation: BlockRotation): BlockState = state.with(FACING, rotation.rotate(state.get(FACING)))

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        builder.add(FACING)
    }

    override fun canPathfindThrough(state: BlockState, world: BlockView, pos: BlockPos, type: NavigationType) = false

    override fun getColor(state: BlockState, world: BlockView, pos: BlockPos) = state.getMapColor(world, pos).color

    companion object {
        val CODEC: MapCodec<NetheriteAnvilBlock> = createCodec { settings: Settings -> NetheriteAnvilBlock(settings) }

        val FACING: DirectionProperty = HorizontalFacingBlock.FACING
        private val BASE_SHAPE = createCuboidShape(2.0, 0.0, 2.0, 14.0, 4.0, 14.0)
        private val X_STEP_SHAPE = createCuboidShape(3.0, 4.0, 4.0, 13.0, 5.0, 12.0)
        private val X_STEM_SHAPE = createCuboidShape(4.0, 5.0, 6.0, 12.0, 10.0, 10.0)
        private val X_FACE_SHAPE = createCuboidShape(0.0, 10.0, 3.0, 16.0, 16.0, 13.0)
        private val Z_STEP_SHAPE = createCuboidShape(4.0, 4.0, 3.0, 12.0, 5.0, 13.0)
        private val Z_STEM_SHAPE = createCuboidShape(6.0, 5.0, 4.0, 10.0, 10.0, 12.0)
        private val Z_FACE_SHAPE = createCuboidShape(3.0, 10.0, 0.0, 13.0, 16.0, 16.0)
        private val X_AXIS_SHAPE = VoxelShapes.union(BASE_SHAPE, X_STEP_SHAPE, X_STEM_SHAPE, X_FACE_SHAPE)
        private val Z_AXIS_SHAPE = VoxelShapes.union(BASE_SHAPE, Z_STEP_SHAPE, Z_STEM_SHAPE, Z_FACE_SHAPE)
        private val TITLE = Text.translatable("container.repair")
        private const val FALLING_BLOCK_ENTITY_DAMAGE_MULTIPLIER = 2.0f
        private const val FALLING_BLOCK_ENTITY_MAX_DAMAGE = 40

        // TODO: handle damaging of anvil
        fun getLandingState(fallingState: BlockState): BlockState? {
            return when {
                fallingState.isOf(BlockRegistry.NETHERITE_ANVIL) -> {
                    BlockRegistry.CHIPPED_NETHERITE_ANVIL.defaultState.with(FACING, fallingState.get(FACING)) as BlockState
                }

                fallingState.isOf(BlockRegistry.CHIPPED_NETHERITE_ANVIL) -> {
                    BlockRegistry.DAMAGED_NETHERITE_ANVIL.defaultState.with(FACING, fallingState.get(FACING)) as BlockState
                }

                else -> null
            }
        }
    }
}

